code/datums/components/storage/ui.dm

```diff
@@ -0,0 +1,285 @@
+		I.maptext = ""
+		I.layer = VOLUMETRIC_STORAGE_ITEM_LAYER
+		I.plane = VOLUMETRIC_STORAGE_ITEM_PLANE
+
+		// finally add our things.
+		. += B.on_screen_objects()
+		. += I
+
+	// Then, continuous section.
+	ui_continuous = get_ui_continuous()
+	ui_continuous.screen_loc = "[screen_start_x]:[screen_pixel_x],[screen_start_y]:[screen_pixel_y] to [screen_start_x+maxcolumns-1]:[screen_pixel_x],[screen_start_y]:[screen_pixel_y]"
+	. += ui_continuous
+	// Then, left and right.
+	ui_left = get_ui_left()
+	ui_left.screen_loc = "[screen_start_x]:[screen_pixel_x - 2],[screen_start_y]:[screen_pixel_y]"
+	. += ui_left
+	ui_right = get_ui_right()
+	ui_right.screen_loc = "[screen_start_x + maxcolumns]:[screen_pixel_x],[screen_start_y]:[screen_pixel_y]"
+	. += ui_right
+	// Then, closer.
+	ui_close = get_ui_close()
+	ui_close.screen_loc = "[screen_start_x + maxcolumns]:[screen_pixel_x + 2],[screen_start_y]:[screen_pixel_y]"
+	. += ui_close
+
+/**
+  * Shows our UI to a mob.
+  */
+/datum/component/storage/proc/ui_show(mob/M, set_screen_size = TRUE)
+	if(!M.client)
+		return FALSE
+	var/list/cview = getviewsize(M.client.view)
+	// in tiles
+	var/maxallowedscreensize = cview[1]-8
+	if(set_screen_size)
+		current_maxscreensize = maxallowedscreensize
+	else if(current_maxscreensize)
+		maxallowedscreensize = current_maxscreensize
+	// we got screen size, register signal
+	RegisterSignal(M, COMSIG_MOB_CLIENT_LOGOUT, .proc/on_logout, override = TRUE)
+	if(M.active_storage != src)
+		if(M.active_storage)
+			M.active_storage.ui_hide(M)
+		M.active_storage = src
+	LAZYOR(is_using, M)
+	if(volumetric_ui())
+		//new volumetric ui bay-style
+		M.client.screen |= orient2hud_volumetric(M, maxallowedscreensize)
+	else
+		//old ui
+		M.client.screen |= orient2hud_legacy(M, maxallowedscreensize)
+	return TRUE
+
+/**
+  * VV hooked to ensure no lingering screen objects.
+  */
+/datum/component/storage/vv_edit_var(var_name, var_value)
+	var/list/old
+	if(var_name == NAMEOF(src, storage_flags))
+		old = is_using.Copy()
+		for(var/i in is_using)
+			ui_hide(i)
+	. = ..()
+	if(old)
+		for(var/i in old)
+			ui_show(i)
+
+/**
+  * Proc triggered by signal to ensure logging out clients don't linger.
+  */
+/datum/component/storage/proc/on_logout(datum/source, client/C)
+	ui_hide(source)
+
+/**
+  * Hides our UI from a mob
+  */
+/datum/component/storage/proc/ui_hide(mob/M)
+	if(!M.client)
+		return TRUE
+	UnregisterSignal(M, COMSIG_MOB_CLIENT_LOGOUT)
+	M.client.screen -= list(ui_boxes, ui_close, ui_left, ui_right, ui_continuous) + get_ui_item_objects_hide()
+	if(M.active_storage == src)
+		M.active_storage = null
+	LAZYREMOVE(is_using, M)
+	return TRUE
+
+/**
+  * Returns TRUE if we are using volumetric UI instead of box UI
+  */
+/datum/component/storage/proc/volumetric_ui()
+	var/atom/real_location = real_location()
+	return (storage_flags & STORAGE_LIMIT_VOLUME) && (length(real_location.contents) <= MAXIMUM_VOLUMETRIC_ITEMS) && !display_numerical_stacking
+
+/**
```

Comment:
anyone wondering why i'm spamming getters:

yeah it's pretty terrible to have every storage component making 2 screen objects much less 5 even if no player uses it